operator/pkg/helmreconciler/reconcile.go

```diff
@@ -0,0 +1,336 @@
+		return err
+	}
+
+	status := h.processRecursive(manifestMap)
+
+	// Delete any resources not in the manifest but managed by operator.
+	var errs util.Errors
+	if h.needUpdateAndPrune {
+		errs = util.AppendErr(errs, h.Prune(allObjectHashes(manifestMap), false))
+	}
+	errs = util.AppendErr(errs, h.endReconcile(status))
+
+	return errs.ToError()
+}
+
+// beginReconcile updates the status field on the IstioOperator instance before reconciling.
+func (h *HelmReconciler) beginReconcile() error {
+	if !h.opts.ControllerMode || h.opts.DryRun {
+		return nil
+	}
+	isop := &valuesv1alpha1.IstioOperator{}
+	namespacedName := types.NamespacedName{
+		Name:      h.iop.Name,
+		Namespace: h.iop.Namespace,
+	}
+	if err := h.GetClient().Get(context.TODO(), namespacedName, isop); err != nil {
+		if runtime.IsNotRegisteredError(err) {
+			// CRD not yet installed in cluster, nothing to update.
+			return nil
+		}
+		return fmt.Errorf("failed to get IstioOperator before updating status due to %v", err)
+	}
+	if isop.Status == nil {
+		isop.Status = &v1alpha1.InstallStatus{Status: v1alpha1.InstallStatus_RECONCILING}
+	} else {
+		cs := isop.Status.ComponentStatus
+		for cn := range cs {
+			cs[cn] = &v1alpha1.InstallStatus_VersionStatus{
+				Status: v1alpha1.InstallStatus_RECONCILING,
+			}
+		}
+		isop.Status.Status = v1alpha1.InstallStatus_RECONCILING
+	}
+	return h.GetClient().Status().Update(context.TODO(), isop)
+}
+
+// endReconcile updates the status field on the IstioOperator instance based on the resulting err parameter.
+func (h *HelmReconciler) endReconcile(status *v1alpha1.InstallStatus) error {
+	if !h.opts.ControllerMode || h.opts.DryRun {
+		return nil
+	}
+	iop := &valuesv1alpha1.IstioOperator{}
+	namespacedName := types.NamespacedName{
+		Name:      h.iop.Name,
+		Namespace: h.iop.Namespace,
+	}
+	if err := h.GetClient().Get(context.TODO(), namespacedName, iop); err != nil {
+		return fmt.Errorf("failed to get IstioOperator before updating status due to %v", err)
+	}
+	iop.Status = status
+	return h.GetClient().Status().Update(context.TODO(), iop)
+}
+
+// processRecursive processes the given manifests in an order of dependencies defined in h. Dependencies are a tree,
+// where a child must wait for the parent to complete before starting.
+func (h *HelmReconciler) processRecursive(manifests ChartManifestsMap) *v1alpha1.InstallStatus {
+	componentStatus := make(map[string]*v1alpha1.InstallStatus_VersionStatus)
+
+	// mu protects the shared InstallStatus componentStatus across goroutines
+	var mu sync.Mutex
+	// wg waits for all manifest processing goroutines to finish
+	var wg sync.WaitGroup
+
+	for c, m := range manifests {
+		c, m := c, m
+		wg.Add(1)
+		go func() {
+			var processedObjs object.K8sObjects
+			defer wg.Done()
+			cn := name.ComponentName(c)
+			if s := dependencyWaitCh[cn]; s != nil {
+				log.Infof("%s is waiting on dependency...", c)
+				<-s
+				log.Infof("Dependency for %s has completed, proceeding.", c)
+			}
```

Comment:
not: I found that the default timeout for component dependency waiting is **10** mins. So before timeout, even if we manually delete IOP instance, it will be still stuck in waiting period until timeout, then the next reconcile loop(iop deleted) start. This may be confused, since overall the reconcile loop is single routine.

Reply:
thanks @morvencao - the timeout is experimentally set to work with integration tests but i agree it's probably too long. separately from this pr it may be a good idea to expose this timeout in some way (env var maybe?)

Reply:
Agree to have a separate PR to fix that.